File: js/src/fingerprint/init-script.js
/**
 * Page init script for the fingerprint surfaces CDP cannot override.
 *
 * Everything here is strictly worse than a browser-enforced override: a
 * JavaScript patch is visible to anyone who inspects the property descriptor
 * carefully enough, and it does not reach workers or HTTP headers. It exists
 * only for fields the `Emulation` domain has no command for -- see
 * `limitations.js` for the measured list -- and for `connectBrowser`, where the
 * launch switches were already fixed by whoever started the browser.
 *
 * The payload is written as a real function and serialized with `String()`
 * rather than kept in a template literal, so it stays lintable and formattable
 * like the rest of the source.
 */

/**
 * Runs inside the page, before any page script, in every frame.
 *
 * @param {object} config Plain JSON produced by `buildInitScriptConfig`.
 */
function fingerprintPayload(config) {
  const scope = globalThis;
  // Playwright and Puppeteer both re-inject init scripts on same-document
  // navigations in some versions; patching twice would stack getters and make
  // the descriptor chain observable.
  const marker = '__browserCommanderFingerprint';
  if (scope[marker]) {
    return;
  }
  Object.defineProperty(scope, marker, {
    value: true,
    configurable: true,
    enumerable: false,
    writable: false,
  });

  const nativeSources = new WeakMap();
  const functionToString = Function.prototype.toString;
  const objectDefineProperty = Object.defineProperty;

  /** Make a replacement report `[native code]`, like the accessor it replaces. */
  function asNative(fn, name) {
    try {
      objectDefineProperty(fn, 'name', {
        value: name,
        configurable: true,
        enumerable: false,
        writable: false,
      });
    } catch {
      // A frozen function name is not worth failing the whole patch over.
    }
    nativeSources.set(fn, `function ${name}() { [native code] }`);
    return fn;
  }

  const patchedToString = function toString() {
    const source = nativeSources.get(this);
    return source === undefined ? functionToString.call(this) : source;
  };
  nativeSources.set(patchedToString, functionToString.call(functionToString));
  Function.prototype.toString = patchedToString;

  /**
   * Replace a native accessor on a prototype, keeping the shape a real one has:
   * an enumerable, configurable getter with no setter, named `get <property>`.
   */
  function defineNativeGetter(target, property, produce) {
    if (!target) {
      return;
    }
    const descriptor = Object.getOwnPropertyDescriptor(target, property);
    if (descriptor && !descriptor.configurable) {
      return;
    }
    const getter = asNative(() => produce(), `get ${property}`);
    try {
      objectDefineProperty(target, property, {
        get: getter,
        set: undefined,
        enumerable: descriptor ? descriptor.enumerable : true,
        configurable: true,
      });
    } catch {
      // Some engines refuse redefinition on cross-origin-ish prototypes.
    }
  }

  const navigatorPrototype = scope.Navigator && scope.Navigator.prototype;
  const workerNavigatorPrototype =
    scope.WorkerNavigator && scope.WorkerNavigator.prototype;

  function defineNavigator(property, value) {
    defineNativeGetter(navigatorPrototype, property, () => value);
    defineNativeGetter(workerNavigatorPrototype, property, () => value);
  }

  if (config.webdriver !== undefined) {
    defineNavigator('webdriver', config.webdriver);
  }
  if (config.deviceMemory !== undefined) {
    defineNavigator('deviceMemory', config.deviceMemory);
  }
  if (config.hardwareConcurrency !== undefined) {
    defineNavigator('hardwareConcurrency', config.hardwareConcurrency);
  }
  if (config.vendor !== undefined) {
    defineNavigator('vendor', config.vendor);
  }
  if (config.platform !== undefined) {
    defineNavigator('platform', config.platform);
  }
  if (config.maxTouchPoints !== undefined) {
    defineNavigator('maxTouchPoints', config.maxTouchPoints);
  }
  if (config.doNotTrack !== undefined) {
    defineNativeGetter(
      navigatorPrototype,
      'doNotTrack',
      () => config.doNotTrack
    );
  }
  if (config.languages !== undefined) {
    // navigator.languages is a frozen array in Chrome; hand out a fresh frozen
    // copy so page code cannot mutate the one we hold.
    defineNavigator('language', config.languages[0]);
    defineNativeGetter(navigatorPrototype, 'languages', () =>
      Object.freeze(config.languages.slice())
    );
    defineNativeGetter(workerNavigatorPrototype, 'languages', () =>
      Object.freeze(config.languages.slice())
    );
  }

  const screenPrototype = scope.Screen && scope.Screen.prototype;
  if (config.screen) {
    for (const property of Object.keys(config.screen)) {
      const value = config.screen[property];
      defineNativeGetter(screenPrototype, property, () => value);
    }
  }

  if (config.webgl) {
    // 0x9245/0x9246 are UNMASKED_VENDOR_WEBGL and UNMASKED_RENDERER_WEBGL from
    // the WEBGL_debug_renderer_info extension; 0x1F00/0x1F01 are the plain
    // VENDOR and RENDERER constants that need no extension at all.
    const replacements = new Map();
    if (config.webgl.unmaskedVendor !== undefined) {
      replacements.set(0x9245, config.webgl.unmaskedVendor);
    }
    if (config.webgl.unmaskedRenderer !== undefined) {
      replacements.set(0x9246, config.webgl.unmaskedRenderer);
    }
    if (config.webgl.vendor !== undefined) {
      replacements.set(0x1f00, config.webgl.vendor);
    }
    if (config.webgl.renderer !== undefined) {
      replacements.set(0x1f01, config.webgl.renderer);
    }
    for (const name of ['WebGLRenderingContext', 'WebGL2RenderingContext']) {
      const prototype = scope[name] && scope[name].prototype;
      if (!prototype || typeof prototype.getParameter !== 'function') {
        continue;
      }
      const original = prototype.getParameter;
      const patched = asNative(function getParameter(parameter) {
        if (replacements.has(parameter)) {
          return replacements.get(parameter);
        }
        return original.call(this, parameter);
      }, 'getParameter');
      prototype.getParameter = patched;
    }
  }
}

const SCREEN_FIELDS = [
  'width',
  'height',
  'availWidth',
  'availHeight',
  'colorDepth',
  'pixelDepth',
];

/**
 * Decide what the init script still has to do once the browser-side overrides
 * have been applied.
 *
 * @param {object} profile Normalized fingerprint profile.
 * @param {object} [options]
 * @param {boolean} [options.patchWebdriver] Force `navigator.webdriver` to
 *   `false` from JavaScript. Only needed when the browser was launched by
 *   somebody else and the automation switches can no longer be changed.
 * @param {boolean} [options.patchLanguages] Also patch `navigator.languages`.
 * @returns {object|null} Config for the payload, or `null` when nothing is left.
 */
export function buildInitScriptConfig(profile, options = {}) {
  const config = {};

  if (options.patchWebdriver) {
    config.webdriver = false;
  }
  if (profile.deviceMemory !== undefined) {
    config.deviceMemory = profile.deviceMemory;
  }
  if (profile.vendor !== undefined) {
    config.vendor = profile.vendor;
  }
  if (profile.doNotTrack !== undefined) {
    config.doNotTrack = profile.doNotTrack;
  }
  if (options.patchLanguages && profile.languages !== undefined) {
    config.languages = [...profile.languages];
  }
  if (profile.webgl !== undefined) {
    config.webgl = { ...profile.webgl };
  }
  if (profile.screen !== undefined) {
    const screen = {};
    for (const field of SCREEN_FIELDS) {
      if (profile.screen[field] !== undefined) {
        screen[field] = profile.screen[field];
      }
    }
    // width and height are already enforced by setDeviceMetricsOverride; the
    // avail*/depth fields are not, so only those need patching.
    delete screen.width;
    delete screen.height;
    if (Object.keys(screen).length > 0) {
      config.screen = screen;
    }
  }

  return Object.keys(config).length > 0 ? config : null;
}

/**
 * Serialize the init script for a profile.
 *
 * @param {object} profile Normalized fingerprint profile.
 * @param {object} [options] See `buildInitScriptConfig`.
 * @returns {string|null} Script source, or `null` when no patching is needed.
 */
export function buildFingerprintInitScript(profile, options = {}) {
  if (!profile || typeof profile !== 'object') {
    throw new TypeError('profile must be a normalized fingerprint profile');
  }
  const config = buildInitScriptConfig(profile, options);
  if (!config) {
    return null;
  }
  return `(${String(fingerprintPayload)})(${JSON.stringify(config)});`;
}

File: rust/src/fingerprint/init_script.rs
//! Page init script for the fingerprint surfaces CDP cannot override.
//!
//! Everything here is strictly worse than a browser-enforced override: a
//! JavaScript patch is visible to anyone who inspects the property descriptor
//! carefully enough, and it does not reach workers or HTTP headers. It exists
//! only for fields the `Emulation` domain has no command for, and for
//! connecting to a browser somebody else launched, where the switches can no
//! longer be changed.
//!
//! The payload itself is not written here. `init_payload.js` next to this
//! module is a byte-for-byte copy of `js/src/fingerprint/init-payload.js`, kept
//! in step by `scripts/check-shared-init-payload.sh`, so all three
//! implementations send Chrome the same script rather than three hand-written
//! translations of it.

use serde_json::{json, Map, Value};

use super::profile::FingerprintProfile;

/// The shared payload source, embedded at compile time.
pub const FINGERPRINT_PAYLOAD_SOURCE: &str = include_str!("init_payload.js");

/// What the init script still has to patch after the CDP overrides.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct InitScriptOptions {
    /// Force `navigator.webdriver` to `false` from JavaScript.
    ///
    /// Only needed when the browser was launched by somebody else and
    /// `--disable-blink-features=AutomationControlled` can no longer be passed.
    pub patch_webdriver: bool,
    /// Also patch `navigator.languages`, which the browser already sets from
    /// `acceptLanguage`.
    pub patch_languages: bool,
}

/// Decide what the init script still has to do after the CDP overrides.
///
/// Returns `None` when the browser-side overrides already cover everything.
pub fn build_init_script_config(
    profile: &FingerprintProfile,
    options: InitScriptOptions,
) -> Option<Value> {
    let mut config = Map::new();

    if options.patch_webdriver {
        config.insert("webdriver".to_string(), json!(false));
    }
    if let Some(device_memory) = profile.device_memory {
        config.insert("deviceMemory".to_string(), json!(device_memory));
    }
    if let Some(vendor) = &profile.vendor {
        config.insert("vendor".to_string(), json!(vendor));
    }
    if let Some(do_not_track) = &profile.do_not_track {
        config.insert("doNotTrack".to_string(), json!(do_not_track));
    }
    if options.patch_languages {
        if let Some(languages) = &profile.languages {
            config.insert("languages".to_string(), json!(languages));
        }
    }
    if let Some(webgl) = &profile.webgl {
        config.insert("webgl".to_string(), json!(webgl));
    }
    if let Some(screen) = &profile.screen {
        // width and height are already enforced by setDeviceMetricsOverride;
        // the avail*/depth fields are not, so only those need patching.
        let mut patched = Map::new();
        for (name, value) in [
            ("availWidth", screen.avail_width),
            ("availHeight", screen.avail_height),
            ("colorDepth", screen.color_depth),
            ("pixelDepth", screen.pixel_depth),
        ] {
            if let Some(value) = value {
                patched.insert(name.to_string(), json!(value));
            }
        }
        if !patched.is_empty() {
            config.insert("screen".to_string(), Value::Object(patched));
        }
    }

    if config.is_empty() {
        None
    } else {
        Some(Value::Object(config))
    }
}

/// Serialize the init script for a profile, or `None` when none is needed.
pub fn build_fingerprint_init_script(
    profile: &FingerprintProfile,
    options: InitScriptOptions,
) -> Option<String> {
    let config = build_init_script_config(profile, options)?;
    // The payload is wrapped in an IIFE so the declaration never becomes a
    // property of the page's global object; a stray `fingerprintPayload` global
    // would be a far louder signal than anything the payload hides.
    Some(format!(
        "(() => {{\n{FINGERPRINT_PAYLOAD_SOURCE}\nfingerprintPayload({config});\n}})();"
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::fingerprint::profile::{ScreenProfile, WebglProfile};

    fn config(profile: FingerprintProfile, options: InitScriptOptions) -> Option<Value> {
        build_init_script_config(&profile.resolve().expect("profile resolves"), options)
    }

    fn script(profile: FingerprintProfile, options: InitScriptOptions) -> Option<String> {
        build_fingerprint_init_script(&profile.resolve().expect("profile resolves"), options)
    }

    #[test]
    fn returns_nothing_when_the_browser_side_overrides_cover_everything() {
        let profile = FingerprintProfile::default()
            .user_agent("Mozilla/5.0 (X11; Linux x86_64) Chrome/140.0.0.0")
            .timezone_id("UTC")
            .hardware_concurrency(8)
            .max_touch_points(0);

        assert_eq!(config(profile.clone(), InitScriptOptions::default()), None);
        assert_eq!(script(profile, InitScriptOptions::default()), None);
    }

    #[test]
    fn patches_only_the_fields_the_emulation_domain_has_no_command_for() {
        // hardwareConcurrency and timezoneId are browser-enforced, so they must
        // not appear in the weaker JavaScript patch.
        let config = config(
            FingerprintProfile::default()
                .device_memory(8.0)
                .vendor("Google Inc.")
                .do_not_track("1")
                .hardware_concurrency(8)
                .timezone_id("UTC"),
            InitScriptOptions::default(),
        )
        .expect("something is left to patch");
        let mut keys: Vec<_> = config
            .as_object()
            .expect("object")
            .keys()
            .map(String::as_str)
            .collect();
        keys.sort_unstable();

        assert_eq!(keys, vec!["deviceMemory", "doNotTrack", "vendor"]);
    }

    #[test]
    fn adds_webdriver_only_when_the_caller_asks_for_it() {
        let profile = FingerprintProfile::default().vendor("Google Inc.");

        assert_eq!(
            config(profile.clone(), InitScriptOptions::default()).expect("config")["webdriver"],
            Value::Null
        );
        assert_eq!(
            config(
                profile,
                InitScriptOptions {
                    patch_webdriver: true,
                    ..InitScriptOptions::default()
                }
            )
            .expect("config")["webdriver"],
            json!(false)
        );
    }

    #[test]
    fn patches_webdriver_even_for_an_otherwise_empty_profile() {
        let script = script(
            FingerprintProfile::default(),
            InitScriptOptions {
                patch_webdriver: true,
                ..InitScriptOptions::default()
            },
        )
        .expect("a script is produced");

        assert!(script.contains("\"webdriver\":false"));
    }

    #[test]
    fn leaves_languages_to_the_browser_unless_explicitly_asked() {
        let profile = FingerprintProfile::default().languages(["fr-FR", "fr"]);

        assert_eq!(config(profile.clone(), InitScriptOptions::default()), None);
        assert_eq!(
            config(
                profile,
                InitScriptOptions {
                    patch_languages: true,
                    ..InitScriptOptions::default()
                }
            )
            .expect("config")["languages"],
            json!(["fr-FR", "fr"])
        );
    }

    #[test]
    fn drops_the_screen_dimensions_set_device_metrics_override_already_enforces() {
        let config = config(
            FingerprintProfile::default().screen(ScreenProfile {
                width: Some(1920),
                height: Some(1080),
                avail_width: Some(1920),
                avail_height: Some(1032),
                color_depth: Some(24),
                pixel_depth: Some(24),
            }),
            InitScriptOptions::default(),
        )
        .expect("config");

        assert_eq!(
            config["screen"],
            json!({
                "availWidth": 1920,
                "availHeight": 1032,
                "colorDepth": 24,
                "pixelDepth": 24,
            })
        );
    }

    #[test]
    fn skips_the_screen_patch_when_only_width_and_height_are_given() {
        assert_eq!(
            config(
                FingerprintProfile::default().screen(ScreenProfile {
                    width: Some(1920),
                    height: Some(1080),
                    ..ScreenProfile::default()
                }),
                InitScriptOptions::default()
            ),
            None
        );
    }

    #[test]
    fn wraps_the_shared_payload_and_calls_it_with_the_config() {
        let script = script(
            FingerprintProfile::default().webgl(WebglProfile {
                unmasked_vendor: Some("Google Inc. (NVIDIA)".to_string()),
                ..WebglProfile::default()
            }),
            InitScriptOptions::default(),
        )
        .expect("a script is produced");

        assert!(script.starts_with("(() => {\n"));
        assert!(script.ends_with("\n})();"));
        assert!(script.contains(FINGERPRINT_PAYLOAD_SOURCE));
        assert!(script.contains("fingerprintPayload({"));
        assert!(script.contains("Google Inc. (NVIDIA)"));
    }

    // The payload is one asset shared with the JavaScript and Python packages;
    // it is source text for a classic script, so module syntax would be a
    // syntax error in the page rather than a failure here.
    #[test]
    fn embeds_the_shared_payload_asset_verbatim() {
        assert!(FINGERPRINT_PAYLOAD_SOURCE.contains("function fingerprintPayload(config) {"));
        for line in FINGERPRINT_PAYLOAD_SOURCE.lines() {
            assert!(
                !line.starts_with("import ") && !line.starts_with("export "),
                "the payload must stay a classic script: {line}"
            );
        }
    }
}
